### PR TITLE
help petr navigate this mess

### PR DESCRIPTION
**Pull request by maiqueb**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,70 @@
...
+    "mode": "bridge",
+    "mtu": 1500,
+    "mac": "02:03:04:05:06:07"
+}
+```
+
+## Network Configuration Reference
+
+* `name`   (string, required): the name of the network.
+* `type`   (string, required): "ovs".
```
Comment (phoracek) [inline README.md]: macvtap

File: README.md [hunk 2]
```diff
@@ -1,2 +1,70 @@
...
+## Example Configuration
+
+```json
+{
+    "name": "mynet",
+    "type": "macvtap",
+    "master": "eth0",
+    "mode": "bridge",
+    "mtu": 1500,
+    "mac": "02:03:04:05:06:07"
```
Comment (phoracek) [inline README.md]: This should go through env variable, not through configuration. Remember, configuration is set in NetworkAttachmentDefinition and shared by all attachments.

File: macvtap.go [hunk 3]
```diff
@@ -0,0 +1,266 @@
+// Copyright 2015 CNI authors
```
Comment (phoracek) [inline macvtap.go]: Ancient!

File: macvtap.go [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+			return nil
+		})
+
+		if err != nil {
+			return err
+		}
+	}
+
+	// Delete link if err to avoid link leak in this ns
+	defer func() {
```
Comment (phoracek) [inline macvtap.go]: Why is calling this at the end of the function useful? There is nothing that could fail after this point.

File: macvtap.go [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+	}
+
+	m, err := netlink.LinkByName(conf.Master)
+	if err != nil {
+		return nil, fmt.Errorf("failed to lookup master %q: %v", conf.Master, err)
+	}
+
+	// due to kernel bug we have to create with tmpName or it might
+	// collide with the name on the host and error out
+	tmpName, err := ip.RandomVethName()
```
Comment (phoracek) [inline macvtap.go:141]: With this you will create a macvtap device with "veth" in its name. I would not want to debug that.

File: macvtap.go [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+			Mode: mode,
+		},
+	}
+
+	if err := netlink.LinkAdd(mv); err != nil {
+		return nil, fmt.Errorf("failed to create macvtap: %v", err)
+	}
+
+	err = netns.Do(func(_ ns.NetNS) error {
+		// TODO: duplicate following lines for ipv6 support, when it will be added in other places
```
Comment (phoracek) [inline macvtap.go:163]: As a random contributor, I would not understand what do you want from me here

File: macvtap.go [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+		ipv4SysctlValueName := fmt.Sprintf(IPv4InterfaceArpProxySysctlTemplate, tmpName)
```
Comment (phoracek) [inline macvtap.go:164]: I would mention that it is an arp config there

File: macvtap.go [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+		if _, err := sysctl.Sysctl(ipv4SysctlValueName, "1"); err != nil {
+			// remove the newly added link and ignore errors, because we already are in a failed state
+			_ = netlink.LinkDel(mv)
```
Comment (phoracek) [inline macvtap.go:167]: You don't need to assign the return value if you don't use it

File: macvtap.go [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+			return fmt.Errorf("failed to set proxy_arp on newly added interface %q: %v", tmpName, err)
+		}
+
+		err := ip.RenameLink(tmpName, ifName)
+		if err != nil {
+			_ = netlink.LinkDel(mv)
```
Comment (phoracek) [inline macvtap.go:173]: ditto

**Review by phoracek — CHANGES_REQUESTED** [review]

File: README.md (hunk 2 above)
Comment (maiqueb) [reply README.md]: Handled on https://github.com/maiqueb/macvtap-cni/pull/2

File: README.md (hunk 1 above)
Comment (maiqueb) [reply README.md]: Done.

File: macvtap.go (hunk 3 above)
Comment (maiqueb) [reply macvtap.go]: Done

File: macvtap.go (hunk 4 above)
Comment (maiqueb) [reply macvtap.go]: Done in https://github.com/maiqueb/macvtap-cni/pull/3

File: macvtap.go (hunk 5 above)
Comment (maiqueb) [reply macvtap.go:141]: It's name if moved back to the original in https://github.com/maiqueb/macvtap-cni/pull/1/files#diff-be3d8974c719c4ceced4dedbc3ef5fa7R171 .

**Pull request closed without merging**